File: FINALS SCRIPTS/Unreal/CustomEngineImporter.cpp
#include "CustomEngineImporter.h"
#include "Engine/World.h"
#include "Engine/StaticMeshActor.h"
#include "Components/StaticMeshComponent.h"
#include "Materials/MaterialInstanceDynamic.h"
#include "Engine/StaticMesh.h"
#include "UObject/ConstructorHelpers.h"
#include "Engine/Engine.h"
#include "Dom/JsonObject.h"
#include "Serialization/JsonSerializer.h"
#include "Serialization/JsonReader.h"
#include "HAL/PlatformFilemanager.h"
#include "Misc/FileHelper.h"
#include "Misc/DateTime.h"

UCustomEngineImporter::UCustomEngineImporter()
{
    jsonFilePath = TEXT("");
    logOutput = TEXT("");
}

UCustomEngineImporter::~UCustomEngineImporter()
{
    createdEntities.Empty();
}

//* ╔═════════════════════╗
//* ║ Non - Monobehaviour ║
//* ╚═════════════════════╝

bool UCustomEngineImporter::ImportSceneFromFile(const FString& FilePath)
{
    FString JsonContent = GetJsonContent(FilePath);
    if (JsonContent.IsEmpty())
    {
        LogMessage(TEXT("Error: No JSON content found. Please provide a valid file path."));
        return false;
    }

    return ImportScene(JsonContent);
}

bool UCustomEngineImporter::ImportSceneFromString(const FString& JsonContent)
{
    if (JsonContent.IsEmpty())
    {
        LogMessage(TEXT("Error: No JSON content provided."));
        return false;
    }

    return ImportScene(JsonContent);
}

bool UCustomEngineImporter::ImportScene(const FString& JsonContent)
{
    FSceneData SceneData;
    if (!ParseJsonToSceneData(JsonContent, SceneData))
    {
        LogMessage(TEXT("Error: Failed to parse JSON data."));
        return false;
    }

    LogMessage(FString::Printf(TEXT("Importing scene: %s (Version: %s)"), 
        *SceneData.scene_name, *SceneData.version));

    UWorld* World = GEngine->GetWorldFromContextObject(this, EGetWorldErrorMode::LogAndReturnNull);
    if (!World)
    {
        LogMessage(TEXT("Error: Could not get world context."));
        return false;
    }

    AActor* SceneRoot = World->SpawnActor<AActor>();
    SceneRoot->SetActorLabel(FString::Printf(TEXT("Imported_Scene_%s"), *SceneData.scene_name));
    
    createdEntities.Empty();

    if (SceneData.entities.Num() > 0)
    {
        for (const FEntityData& EntityData : SceneData.entities)
        {
            AActor* Entity = CreateEntityWithoutTransform(EntityData, SceneRoot);
            if (Entity != nullptr)
            {
                createdEntities.Add(EntityData.name, Entity);
            }
        }

        for (const FEntityData& EntityData : SceneData.entities)
        {
            if (!EntityData.parent.IsEmpty() && 
                createdEntities.Contains(EntityData.name) && 
                createdEntities.Contains(EntityData.parent))
            {
                AActor* ChildActor = createdEntities[EntityData.name];
                AActor* ParentActor = createdEntities[EntityData.parent];
                
                ChildActor->AttachToActor(ParentActor, FAttachmentTransformRules::KeepWorldTransform);
                LogMessage(FString::Printf(TEXT("Set parent: %s -> %s"), 
                    *EntityData.name, *EntityData.parent));
            }
        }

        for (const FEntityData& EntityData : SceneData.entities)
        {
            if (createdEntities.Contains(EntityData.name))
            {
                ApplyTransform(createdEntities[EntityData.name], EntityData.transform);
            }
        }
    }

    LogMessage(FString::Printf(TEXT("Successfully imported %d entities from scene '%s'"), 
        createdEntities.Num(), *SceneData.scene_name));
    
    return true;
}

AActor* UCustomEngineImporter::CreateEntityWithoutTransform(const FEntityData& EntityData, AActor* ParentActor)
{
    UWorld* World = GEngine->GetWorldFromContextObject(this, EGetWorldErrorMode::LogAndReturnNull);
    if (!World)
    {
        return nullptr;
    }

    AActor* Entity = World->SpawnActor<AActor>();
    Entity->SetActorLabel(EntityData.name);
    Entity->SetActorHiddenInGame(!EntityData.enabled);

    if (ParentActor)
    {
        Entity->AttachToActor(ParentActor, FAttachmentTransformRules::KeepWorldTransform);
    }

    if (EntityData.components.Num() > 0)
    {
        for (const FComponentData& ComponentData : EntityData.components)
        {
            CreateComponent(Entity, ComponentData, EntityData.constant);
        }
    }

    LogMessage(FString::Printf(TEXT("Created entity: %s"), *EntityData.name));
    return Entity;
}

void UCustomEngineImporter::ApplyTransform(AActor* Entity, const FTransformData& TransformData)
{
    if (!Entity)
    {
        return;
    }

    FTransform NewTransform;
    NewTransform.SetLocation(TransformData.position.ToUnrealVector());
    NewTransform.SetRotation(FQuat::MakeFromEuler(TransformData.rotation.ToUnrealVector()));
    NewTransform.SetScale3D(TransformData.scale.ToUnrealVector());

    Entity->SetActorRelativeTransform(NewTransform);

    LogMessage(FString::Printf(TEXT("Applied transform to %s: pos=%s, rot=%s, scale=%s"),
        *Entity->GetActorLabel(),
        *TransformData.position.ToUnrealVector().ToString(),
        *TransformData.rotation.ToUnrealVector().ToString(),
        *TransformData.scale.ToUnrealVector().ToString()));
}

void UCustomEngineImporter::CreateComponent(AActor* Entity, const FComponentData& ComponentData, const FConstantData& ConstantData)
{
    if (!Entity)
    {
        return;
    }

    UStaticMesh* PrimitiveMesh = nullptr;

    if (ComponentData.type == TEXT("CubeRenderer"))
    {
        PrimitiveMesh = GetPrimitiveMesh(TEXT("Cube"));
    }
    else if (ComponentData.type == TEXT("QuadRenderer"))
    {
        PrimitiveMesh = GetPrimitiveMesh(TEXT("Plane"));
    }
    else if (ComponentData.type == TEXT("SphereRenderer"))
    {
        PrimitiveMesh = GetPrimitiveMesh(TEXT("Sphere"));
    }
    else if (ComponentData.type == TEXT("CylinderRenderer"))
    {
        PrimitiveMesh = GetPrimitiveMesh(TEXT("Cylinder"));
    }
    else if (ComponentData.type == TEXT("MeshRenderer"))
    {
        PrimitiveMesh = nullptr;
        LogMessage(FString::Printf(TEXT("MeshRenderer created for model: %s"), *ComponentData.modelPath));
    }
    else if (ComponentData.type == TEXT("PhysicsComponent"))
    {
        UStaticMeshComponent* MeshComponent = Entity->FindComponentByClass<UStaticMeshComponent>();
        if (MeshComponent)
        {
            MeshComponent->SetCollisionEnabled(ECollisionEnabled::QueryAndPhysics);
            MeshComponent->SetCollisionResponseToAllChannels(ECR_Block);
            
            if (ComponentData.bodyType == 0)
            {
                MeshComponent->SetMobility(EComponentMobility::Static);
            }
            else
            {
                MeshComponent->SetMobility(EComponentMobility::Movable);
                MeshComponent->SetSimulatePhysics(true);
                MeshComponent->SetMassOverrideInKg(NAME_None, ComponentData.mass > 0.0f ? ComponentData.mass : 1.0f);
            }
        }

        LogMessage(FString::Printf(TEXT("Added physics component (mass: %f)"), ComponentData.mass));
        return;
    }

    if (ComponentData.type.Contains(TEXT("Renderer")))
    {
        UStaticMeshComponent* MeshComponent = Entity->FindComponentByClass<UStaticMeshComponent>();
        if (!MeshComponent)
        {
            MeshComponent = NewObject<UStaticMeshComponent>(Entity);
            Entity->AddInstanceComponent(MeshComponent);
            Entity->SetRootComponent(MeshComponent);
        }

        if (PrimitiveMesh != nullptr)
        {
            MeshComponent->SetStaticMesh(PrimitiveMesh);
        }

        if (ConstantData.color.x >= 0.0f && ConstantData.color.y >= 0.0f && ConstantData.color.z >= 0.0f)
        {
            UMaterialInstanceDynamic* DynamicMaterial = CreateMaterialWithColor(ConstantData.color);
            if (DynamicMaterial)
            {
                MeshComponent->SetMaterial(0, DynamicMaterial);
            }
        }

        if (ComponentData.isTextured && !ComponentData.texturePath.IsEmpty())
        {
            LogMessage(FString::Printf(TEXT("Texture path found: %s"), *ComponentData.texturePath));
        }

        LogMessage(FString::Printf(TEXT("Added %s component to entity"), *ComponentData.type));
    }
}

UStaticMesh* UCustomEngineImporter::GetPrimitiveMesh(const FString& PrimitiveType)
{
    UStaticMesh* PrimitiveMesh = nullptr;

    if (PrimitiveType == TEXT("Cube"))
    {
        static ConstructorHelpers::FObjectFinder<UStaticMesh> CubeMeshFinder(TEXT("/Engine/BasicShapes/Cube"));
        PrimitiveMesh = CubeMeshFinder.Object;
    }
    else if (PrimitiveType == TEXT("Sphere"))
    {
        static ConstructorHelpers::FObjectFinder<UStaticMesh> SphereMeshFinder(TEXT("/Engine/BasicShapes/Sphere"));
        PrimitiveMesh = SphereMeshFinder.Object;
    }
    else if (PrimitiveType == TEXT("Cylinder"))
    {
        static ConstructorHelpers::FObjectFinder<UStaticMesh> CylinderMeshFinder(TEXT("/Engine/BasicShapes/Cylinder"));
        PrimitiveMesh = CylinderMeshFinder.Object;
    }
    else if (PrimitiveType == TEXT("Plane"))
    {
        static ConstructorHelpers::FObjectFinder<UStaticMesh> PlaneMeshFinder(TEXT("/Engine/BasicShapes/Plane"));
        PrimitiveMesh = PlaneMeshFinder.Object;
    }

    return PrimitiveMesh;
}

UMaterialInstanceDynamic* UCustomEngineImporter::CreateMaterialWithColor(const FVector3Data& ColorData)
{
    static ConstructorHelpers::FObjectFinder<UMaterial> BaseMaterialFinder(TEXT("/Engine/BasicShapes/BasicShapeMaterial"));
    UMaterial* BaseMaterial = BaseMaterialFinder.Object;
    
    if (!BaseMaterial)
    {
        return nullptr;
    }

    UMaterialInstanceDynamic* DynamicMaterial = UMaterialInstanceDynamic::Create(BaseMaterial, this);
    if (DynamicMaterial)
    {
        FLinearColor Color(ColorData.x, ColorData.y, ColorData.z, 1.0f);
        DynamicMaterial->SetVectorParameterValue(TEXT("Color"), Color);
    }

    return DynamicMaterial;
}

FString UCustomEngineImporter::GetJsonContent(const FString& FilePath)
{
    FString JsonContent;
    
    if (!FilePath.IsEmpty() && FPlatformFileManager::Get().GetPlatformFile().FileExists(*FilePath))
    {
        if (!FFileHelper::LoadFileToString(JsonContent, *FilePath))
        {
            LogMessage(FString::Printf(TEXT("Error: Could not read file: %s"), *FilePath));
        }
    }

    return JsonContent;
}

void UCustomEngineImporter::ClearCurrentScene()
{
    UWorld* World = GEngine->GetWorldFromContextObject(this, EGetWorldErrorMode::LogAndReturnNull);
    if (!World)
    {
        return;
    }

    TArray<AActor*> ActorsToDestroy;
    
    for (TActorIterator<AActor> ActorIterator(World); ActorIterator; ++ActorIterator)
    {
        AActor* CurrentActor = *ActorIterator;
        if (CurrentActor && CurrentActor->GetActorLabel().StartsWith(TEXT("Imported_Scene_")))
        {
            ActorsToDestroy.Add(CurrentActor);
        }
    }

    for (AActor* ActorToDestroy : ActorsToDestroy)
    {
        LogMessage(FString::Printf(TEXT("Cleared imported scene: %s"), *ActorToDestroy->GetActorLabel()));
        ActorToDestroy->Destroy();
    }

    createdEntities.Empty();
}

bool UCustomEngineImporter::ParseJsonToSceneData(const FString& JsonContent, FSceneData& OutSceneData)
{
    TSharedPtr<FJsonObject> JsonObject;
    TSharedRef<TJsonReader<>> Reader = TJsonReaderFactory<>::Create(JsonContent);

    if (!FJsonSerializer::Deserialize(Reader, JsonObject) || !JsonObject.IsValid())
    {
        return false;
    }

    OutSceneData.scene_name = JsonObject->GetStringField(TEXT("scene_name"));
    OutSceneData.version = JsonObject->GetStringField(TEXT("version"));

    const TArray<TSharedPtr<FJsonValue>>* EntitiesArray;
    if (JsonObject->TryGetArrayField(TEXT("entities"), EntitiesArray))
    {
        for (const TSharedPtr<FJsonValue>& EntityValue : *EntitiesArray)
        {
            TSharedPtr<FJsonObject> EntityObject = EntityValue->AsObject();
            if (EntityObject.IsValid())
            {
                FEntityData EntityData;
                EntityData.name = EntityObject->GetStringField(TEXT("name"));
                EntityData.enabled = EntityObject->GetBoolField(TEXT("enabled"));
                EntityData.parent = EntityObject->GetStringField(TEXT("parent"));

                TSharedPtr<FJsonObject> TransformObject = EntityObject->GetObjectField(TEXT("transform"));
                if (TransformObject.IsValid())
                {
                    TSharedPtr<FJsonObject> PositionObject = TransformObject->GetObjectField(TEXT("position"));
                    if (PositionObject.IsValid())
                    {
                        EntityData.transform.position.x = PositionObject->GetNumberField(TEXT("x"));
                        EntityData.transform.position.y = PositionObject->GetNumberField(TEXT("y"));
                        EntityData.transform.position.z = PositionObject->GetNumberField(TEXT("z"));
                    }

                    TSharedPtr<FJsonObject> RotationObject = TransformObject->GetObjectField(TEXT("rotation"));
                    if (RotationObject.IsValid())
                    {
                        EntityData.transform.rotation.x = RotationObject->GetNumberField(TEXT("x"));
                        EntityData.transform.rotation.y = RotationObject->GetNumberField(TEXT("y"));
                        EntityData.transform.rotation.z = RotationObject->GetNumberField(TEXT("z"));
                    }

                    TSharedPtr<FJsonObject> ScaleObject = TransformObject->GetObjectField(TEXT("scale"));
                    if (ScaleObject.IsValid())
                    {
                        EntityData.transform.scale.x = ScaleObject->GetNumberField(TEXT("x"));
                        EntityData.transform.scale.y = ScaleObject->GetNumberField(TEXT("y"));
                        EntityData.transform.scale.z = ScaleObject->GetNumberField(TEXT("z"));
                    }
                }

                TSharedPtr<FJsonObject> ConstantObject = EntityObject->GetObjectField(TEXT("constant"));
                if (ConstantObject.IsValid())
                {
                    TSharedPtr<FJsonObject> ColorObject = ConstantObject->GetObjectField(TEXT("color"));
                    if (ColorObject.IsValid())
                    {
                        EntityData.constant.color.x = ColorObject->GetNumberField(TEXT("x"));
                        EntityData.constant.color.y = ColorObject->GetNumberField(TEXT("y"));
                        EntityData.constant.color.z = ColorObject->GetNumberField(TEXT("z"));
                    }
                    EntityData.constant.angle = ConstantObject->GetNumberField(TEXT("angle"));
                    EntityData.constant.isRandom = ConstantObject->GetIntegerField(TEXT("isRandom"));
                    EntityData.constant.hasTex = ConstantObject->GetIntegerField(TEXT("hasTex"));
                }

                const TArray<TSharedPtr<FJsonValue>>* ComponentsArray;
                if (EntityObject->TryGetArrayField(TEXT("components"), ComponentsArray))
                {
                    for (const TSharedPtr<FJsonValue>& ComponentValue : *ComponentsArray)
                    {
                        TSharedPtr<FJsonObject> ComponentObject = ComponentValue->AsObject();
                        if (ComponentObject.IsValid())
                        {
                            FComponentData ComponentData;
                            ComponentData.type = ComponentObject->GetStringField(TEXT("type"));
                            ComponentData.isTextured = ComponentObject->GetBoolField(TEXT("isTextured"));
                            ComponentData.modelPath = ComponentObject->GetStringField(TEXT("modelPath"));
                            ComponentData.texturePath = ComponentObject->GetStringField(TEXT("texturePath"));
                            ComponentData.bodyType = ComponentObject->GetIntegerField(TEXT("bodyType"));
                            ComponentData.mass = ComponentObject->GetNumberField(TEXT("mass"));

                            EntityData.components.Add(ComponentData);
                        }
                    }
                }

                OutSceneData.entities.Add(EntityData);
            }
        }
    }

    return true;
}

void UCustomEngineImporter::LogMessage(const FString& Message)
{
    FDateTime Now = FDateTime::Now();
    FString TimeStamp = Now.ToString(TEXT("%H:%M:%S"));
    
    logOutput += FString::Printf(TEXT("[%s] %s\n"), *TimeStamp, *Message);
    UE_LOG(LogTemp, Log, TEXT("Custom Engine Importer: %s"), *Message);
}

//* ╔════════════════════════════════╗
//* ║ Virtual / Overridden Functions ║
//* ╚════════════════════════════════╝

File: FINALS SCRIPTS/CustomEngineImporter.cs
using System;
using System.Collections.Generic;
using System.IO;
using UnityEditor;
using UnityEngine;

namespace LilLycanLord_Official
{
    public class CustomEngineImporter : EditorWindow
    {
        //* ╔════════════╗
        //* ║ Components ║
        //* ╚════════════╝

        //* ╔══════════╗
        //* ║ Displays ║
        //* ╚══════════╝

        //* ╔════════╗
        //* ║ Fields ║
        //* ╚════════╝

        private string jsonFilePath = "";
        private TextAsset jsonFile;
        private Vector2 scrollPosition;
        private string logOutput = "";

        //* ╔════════════╗
        //* ║ Attributes ║
        //* ╚════════════╝

        //* ╔═════════════════════════╗
        //* ║ Data Structure Classes  ║
        //* ╚═════════════════════════╝

        [Serializable]
        public class SceneData
        {
            public string scene_name;
            public string version;
            public EntityData[] entities;
        }

        [Serializable]
        public class EntityData
        {
            public string name;
            public bool enabled = true;
            public TransformData transform;
            public ConstantData constant;
            public string parent;
            public ComponentData[] components;
        }

        [Serializable]
        public class TransformData
        {
            public Vector3Data position;
            public Vector3Data rotation;
            public Vector3Data scale;
        }

        [Serializable]
        public class Vector3Data
        {
            public float x,
                y,
                z;

            public Vector3 ToUnityVector3()
            {
                return new Vector3(x, y, z);
            }
        }

        [Serializable]
        public class ConstantData
        {
            public Vector3Data color;
            public float angle;
            public int isRandom;
            public int hasTex;
        }

        [Serializable]
        public class ComponentData
        {
            public string type;
            public bool isTextured;
            public string modelPath;
            public string texturePath;
            public int bodyType;
            public float mass;
        }

        //* ╔═══════════════╗
        //* ║ Editor Window ║
        //* ╚═══════════════╝

        [MenuItem("Tools/Custom Engine Importer")]
        public static void ShowWindow()
        {
            GetWindow<CustomEngineImporter>("Custom Engine Importer");
        }

        void OnGUI()
        {
            GUILayout.Label("Custom Engine JSON Importer", EditorStyles.boldLabel);
            EditorGUILayout.Space();

            GUILayout.Label("Import Method:", EditorStyles.label);

            EditorGUILayout.BeginHorizontal();
            jsonFilePath = EditorGUILayout.TextField("JSON File Path:", jsonFilePath);
            if (GUILayout.Button("Browse", GUILayout.Width(60)))
            {
                string filePath = EditorUtility.OpenFilePanel("Select JSON File", "", "json");
                if (!string.IsNullOrEmpty(filePath))
                {
                    jsonFilePath = filePath;
                }
            }
            EditorGUILayout.EndHorizontal();

            jsonFile = (TextAsset)
                EditorGUILayout.ObjectField("JSON TextAsset:", jsonFile, typeof(TextAsset), false);

            EditorGUILayout.Space();

            if (GUILayout.Button("Import Scene", GUILayout.Height(30)))
            {
                ImportScene();
            }

            EditorGUILayout.Space();

            if (GUILayout.Button("Clear Current Scene", GUILayout.Height(25)))
            {
                ClearCurrentScene();
            }

            EditorGUILayout.Space();

            GUILayout.Label("Import Log:", EditorStyles.boldLabel);
            scrollPosition = EditorGUILayout.BeginScrollView(scrollPosition, GUILayout.Height(200));
            EditorGUILayout.TextArea(logOutput, GUILayout.ExpandHeight(true));
            EditorGUILayout.EndScrollView();
        }

        //* ╔═════════════════════╗
        //* ║ Non - Monobehaviour ║
        //* ╚═════════════════════╝

        private void ImportScene()
        {
            string jsonContent = GetJsonContent();
            if (string.IsNullOrEmpty(jsonContent))
            {
                LogMessage(
                    "Error: No JSON content found. Please provide a file path or TextAsset."
                );
                return;
            }

            try
            {
                SceneData sceneData = JsonUtility.FromJson<SceneData>(jsonContent);
                if (sceneData == null)
                {
                    LogMessage("Error: Failed to parse JSON data.");
                    return;
                }

                LogMessage(
                    $"Importing scene: {sceneData.scene_name} (Version: {sceneData.version})"
                );

                GameObject sceneRoot = new GameObject($"Imported_Scene_{sceneData.scene_name}");
                Undo.RegisterCreatedObjectUndo(sceneRoot, "Import Custom Engine Scene");

                Dictionary<string, GameObject> createdEntities =
                    new Dictionary<string, GameObject>();

                if (sceneData.entities != null)
                {
                    foreach (EntityData entityData in sceneData.entities)
                    {
                        GameObject entity = CreateEntityWithoutTransform(
                            entityData,
                            sceneRoot.transform
                        );
                        if (entity != null)
                        {
                            createdEntities[entityData.name] = entity;
                        }
                    }

                    foreach (EntityData entityData in sceneData.entities)
                    {
                        if (
                            !string.IsNullOrEmpty(entityData.parent)
                            && createdEntities.ContainsKey(entityData.name)
                            && createdEntities.ContainsKey(entityData.parent)
                        )
                        {
                            createdEntities[entityData.name]
                                .transform.SetParent(createdEntities[entityData.parent].transform);
                            LogMessage($"Set parent: {entityData.name} -> {entityData.parent}");
                        }
                    }

                    foreach (EntityData entityData in sceneData.entities)
                    {
                        if (createdEntities.ContainsKey(entityData.name))
                        {
                            ApplyTransform(createdEntities[entityData.name], entityData.transform);
                        }
                    }
                }

                LogMessage(
                    $"Successfully imported {createdEntities.Count} entities from scene '{sceneData.scene_name}'"
                );
            }
            catch (Exception e)
            {
                LogMessage($"Error importing scene: {e.Message}");
            }
        }

        private GameObject CreateEntityWithoutTransform(
            EntityData entityData,
            Transform parentTransform
        )
        {
            GameObject entity = new GameObject(entityData.name);
            entity.transform.SetParent(parentTransform);
            entity.SetActive(entityData.enabled);

            if (entityData.components != null)
            {
                foreach (ComponentData componentData in entityData.components)
                {
                    CreateComponent(entity, componentData, entityData.constant);
                }
            }

            Undo.RegisterCreatedObjectUndo(entity, "Create Entity");
            LogMessage($"Created entity: {entityData.name}");
            return entity;
        }

        private void ApplyTransform(GameObject entity, TransformData transformData)
        {
            if (transformData != null)
            {
                if (transformData.position != null)
                    entity.transform.localPosition = transformData.position.ToUnityVector3();
                if (transformData.rotation != null)
                    entity.transform.localEulerAngles = transformData.rotation.ToUnityVector3();
                if (transformData.scale != null)
                    entity.transform.localScale = transformData.scale.ToUnityVector3();

                LogMessage(
                    $"Applied transform to {entity.name}: pos={transformData.position?.ToUnityVector3()}, rot={transformData.rotation?.ToUnityVector3()}, scale={transformData.scale?.ToUnityVector3()}"
                );
            }
        }

        private void CreateComponent(
            GameObject entity,
            ComponentData componentData,
            ConstantData constantData
        )
        {
            Mesh primitiveMesh = null;

            switch (componentData.type)
            {
                case "CubeRenderer":
                    primitiveMesh = GetPrimitiveMesh(PrimitiveType.Cube);
                    break;
                case "QuadRenderer":
                    primitiveMesh = GetPrimitiveMesh(PrimitiveType.Quad);
                    break;
                case "SphereRenderer":
                    primitiveMesh = GetPrimitiveMesh(PrimitiveType.Sphere);
                    break;
                case "CylinderRenderer":
                    primitiveMesh = GetPrimitiveMesh(PrimitiveType.Cylinder);
                    break;
                case "MeshRenderer":
                    // For custom meshes, we'll add components but leave mesh null for now
                    primitiveMesh = null;
                    LogMessage($"MeshRenderer created for model: {componentData.modelPath}");
                    break;
                case "PhysicsComponent":
                    // Add physics components
                    Rigidbody rigidBody = entity.AddComponent<Rigidbody>();
                    if (componentData.bodyType == 0)
                        rigidBody.isKinematic = true;
                    rigidBody.mass = componentData.mass > 0 ? componentData.mass : 1.0f;

                    MeshRenderer existingRenderer = entity.GetComponent<MeshRenderer>();
                    if (existingRenderer == null)
                        entity.AddComponent<BoxCollider>();
                    else
                        entity.AddComponent<MeshCollider>();

                    LogMessage($"Added physics component (mass: {componentData.mass})");
                    return;
            }

            if (componentData.type.Contains("Renderer"))
            {
                MeshRenderer meshRenderer = entity.GetComponent<MeshRenderer>();
                if (meshRenderer == null)
                    meshRenderer = entity.AddComponent<MeshRenderer>();

                MeshFilter meshFilter = entity.GetComponent<MeshFilter>();
                if (meshFilter == null)
                    meshFilter = entity.AddComponent<MeshFilter>();

                if (primitiveMesh != null)
                    meshFilter.mesh = primitiveMesh;

                if (constantData?.color != null)
                {
                    Material material = new Material(Shader.Find("Universal Render Pipeline/Lit"));
                    material.color = new Color(
                        constantData.color.x,
                        constantData.color.y,
                        constantData.color.z,
                        1.0f
                    );
                    meshRenderer.material = material;
                }

                if (componentData.isTextured && !string.IsNullOrEmpty(componentData.texturePath))
                {
                    LogMessage($"Texture path found: {componentData.texturePath}");
                }

                LogMessage($"Added {componentData.type} component to entity");
            }
        }

        private Mesh GetPrimitiveMesh(PrimitiveType primitiveType)
        {
            GameObject temporaryGameObject = GameObject.CreatePrimitive(primitiveType);
            Mesh primitiveMesh = temporaryGameObject.GetComponent<MeshFilter>().sharedMesh;
            DestroyImmediate(temporaryGameObject);
            return primitiveMesh;
        }

        private string GetJsonContent()
        {
            if (jsonFile != null)
            {
                return jsonFile.text;
            }

            if (!string.IsNullOrEmpty(jsonFilePath) && File.Exists(jsonFilePath))
            {
                return File.ReadAllText(jsonFilePath);
            }

            return null;
        }

        private void ClearCurrentScene()
        {
            GameObject[] rootObjects = UnityEngine
                .SceneManagement.SceneManager.GetActiveScene()
                .GetRootGameObjects();

            foreach (GameObject rootObject in rootObjects)
            {
                if (rootObject.name.StartsWith("Imported_Scene_"))
                {
                    Undo.DestroyObjectImmediate(rootObject);
                    LogMessage($"Cleared imported scene: {rootObject.name}");
                }
            }
        }

        private void LogMessage(string message)
        {
            logOutput += $"[{System.DateTime.Now:HH:mm:ss}] {message}\n";
            Debug.Log($"Custom Engine Importer: {message}");
            Repaint();
        }

        //* ╔════════════════════════════════╗
        //* ║ Virtual / Overridden Functions ║
        //* ╚════════════════════════════════╝
    }
}
